go/vt/vtgate/engine/plan.go

```diff
@@ -31,41 +33,59 @@ import (
 // An instruction (aka Primitive) is typically a tree where
 // each node does its part by combining the results of the
 // sub-nodes.
-type Plan struct {
-	Type         sqlparser.StatementType // The type of query we have
-	Original     string                  // Original is the original query.
-	Instructions Primitive               // Instructions contains the instructions needed to fulfil the query.
-	BindVarNeeds *sqlparser.BindVarNeeds // Stores BindVars needed to be provided as part of expression rewriting
-	Warnings     []*query.QueryWarning   // Warnings that need to be yielded every time this query runs
-	TablesUsed   []string                // TablesUsed is the list of tables that this plan will query
-
-	ExecCount    uint64 // Count of times this plan was executed
-	ExecTime     uint64 // Total execution time
-	ShardQueries uint64 // Total number of shard queries
-	RowsReturned uint64 // Total number of rows
-	RowsAffected uint64 // Total number of rows
-	Errors       uint64 // Total number of errors
-}
+type (
+	PlanType int8
 
-// AddStats updates the plan execution statistics
-func (p *Plan) AddStats(execCount uint64, execTime time.Duration, shardQueries, rowsAffected, rowsReturned, errors uint64) {
-	atomic.AddUint64(&p.ExecCount, execCount)
-	atomic.AddUint64(&p.ExecTime, uint64(execTime))
-	atomic.AddUint64(&p.ShardQueries, shardQueries)
-	atomic.AddUint64(&p.RowsAffected, rowsAffected)
-	atomic.AddUint64(&p.RowsReturned, rowsReturned)
-	atomic.AddUint64(&p.Errors, errors)
+	Plan struct {
+		Type         PlanType                // Type of plan
+		QueryType    sqlparser.StatementType // Type of query
+		Original     string                  // Original is the original query.
+		Instructions Primitive               // Instructions contains the instructions needed to fulfil the query.
+		BindVarNeeds *sqlparser.BindVarNeeds // Stores BindVars needed to be provided as part of expression rewriting
+		Warnings     []*query.QueryWarning   // Warnings that need to be yielded every time this query runs
+		TablesUsed   []string                // TablesUsed is the list of tables that this plan will query
+
+		ExecCount    uint64 // Count of times this plan was executed
+		ExecTime     uint64 // Total execution time
+		ShardQueries uint64 // Total number of shard queries
+		RowsReturned uint64 // Total number of rows
+		RowsAffected uint64 // Total number of rows
+		Errors       uint64 // Total number of errors
+	}
+)
+
+const (
+	PlanUnknown PlanType = iota
+	PlanLocal
+	PlanPassthrough
+	PlanMultiShard
+	PlanLookup
+	PlanScatter
+	PlanJoinOp
+	PlanForeignKey
+	PlanComplex
+	PlanOnlineDDL
+	PlanDirectDDL
+	PlanTransaction
+	PlanTopoOp
+)
```

Comment:
It could be good to have a link to the docs here as a comment (https://github.com/vitessio/website/pull/1935/files#diff-4badcfb66b7411dea206a299af9b9f8e3d349d817ad332798f747b0b9a40ad9f), both to help us know what each value mean and also to remind us that modifying this list implies that we need to change the docs.